keanu-project/src/test/java/io/improbable/keanu/vertices/AssertVertexTest.java

```diff
@@ -0,0 +1,132 @@
+package io.improbable.keanu.vertices;
+
+import io.improbable.keanu.DeterministicRule;
+import io.improbable.keanu.algorithms.mcmc.MetropolisHastings;
+import io.improbable.keanu.algorithms.variational.optimizer.KeanuOptimizer;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.bool.BooleanTensor;
+import io.improbable.keanu.vertices.bool.BoolVertex;
+import io.improbable.keanu.vertices.bool.nonprobabilistic.ConstantBoolVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class AssertVertexTest {
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Rule
+    public DeterministicRule rule = new DeterministicRule();
+
+    @Test
+    public void assertThrowsOnFalseConstBool() {
+        thrown.expect(AssertionError.class);
+        ConstantBoolVertex constBool = new ConstantBoolVertex(BooleanTensor.create(false));
+        constBool.assertTrue().eval();
+    }
+
+    @Test
+    public void assertPassesOnTrueConstBool() {
+        ConstantBoolVertex constBool = new ConstantBoolVertex(BooleanTensor.create(true));
+        constBool.assertTrue().eval();
+    }
+
+    @Test
+    public void lazyEvalThrowsOnFalseConstBool() {
+        thrown.expect(AssertionError.class);
+        ConstantBoolVertex constBool = new ConstantBoolVertex(BooleanTensor.create(false));
+        constBool.assertTrue().lazyEval();
+    }
+
+    @Test
+    public void assertPassesOnRVWithTruePredicate() {
+        UniformVertex uniform = new UniformVertex(0, 5);
+        BoolVertex predicate = uniform.lessThan(new ConstantDoubleVertex(new double[]{10}));
+        predicate.assertTrue().eval();
+    }
+
+    @Test
+    public void assertPassesOnRVWithFalsePredicate() {
+        thrown.expect(AssertionError.class);
+        UniformVertex uniform = new UniformVertex(0, 5);
+        BoolVertex predicate = uniform.greaterThan(new ConstantDoubleVertex(new double[]{10}));
+        predicate.assertTrue().eval();
+    }
+
+    @Test
+    public void samplingWithAssertionWorks() {
+        thrown.expect(AssertionError.class);
```

Comment:
I think it would be good to add some tests that ensure inference still works correctly even when we have a graph that contains asserts (ie most of these tests check that the assertion correctly fires - it would be good to have ones where the assertion *isn't* supposed to fire).

Even better if we have a test that checks we can still use a gradient optimiser even when we have asserts?

Reply:
Good points. Added some for ensuring an assertion that shouldn't fire doesn't fire for optimising/sampling. Also added the gradient optimiser check you suggested.